miniwallet/contracts/nft/Mini721.sol

```diff
@@ -0,0 +1,392 @@
+    function contractURI() public view returns (string memory) {
+        return _contractUri;
+    }
+
+    // ------------------
+    // Functions for external (user) minting
+    // ------------------
+
+    function mintMini(uint256 amount) external payable whenSaleActive {
+        require(
+            totalSupply() + amount < maxMiniTokens,
+            "Mini721: Purchase would exceed cap"
+        );
+        require(amount <= maxPerMint, "Mini721: Amount exceeds max per mint");
+        require(
+            mintPrice * amount <= msg.value,
+            "Mini721: Ether value sent is not correct"
+        );
+        uint256 excess = msg.value - (amount * mintPrice);
+        if (excess > 0) {
+            payable(msg.sender).transfer(excess);
+        }
+        _safeMint(msg.sender, amount);
+        emit MiniMint(totalSupply(), amount, msg.sender);
+    }
+
+    function burn(uint256 id) public onlyOwner whenMetadataNotFrozen {
+        HRC721._burn(id);
+        emit MiniBurn(id);
+    }
+
+    function batchBurn(uint256[] memory ids)
+        public
+        onlyOwner
+        whenMetadataNotFrozen
+    {
+        for (uint32 i = 0; i < ids.length; i++) {
+            uint256 id = ids[i];
+            HRC721._burn(id);
+        }
+        emit MiniBatchBurn(ids);
+    }
+
+    // ------------------
+    // Functions for the owner (Mini minting contracts)
+    // ------------------
+
+    function freezeMetadata() external onlyOwner whenMetadataNotFrozen {
+        metadataFrozen = true;
+    }
+
+    function freezeProvenance() external onlyOwner whenProvenanceNotFrozen {
+        provenanceFrozen = true;
+    }
+
+    function toggleSaleState() external onlyOwner {
+        require(
+            (saleIsActive || (offsetValue != 0)),
+            "cannot start sale until airdrop is complete and offset set"
+        );
+        saleIsActive = !saleIsActive;
+        if (saleIsActive && !saleStarted) {
+            saleStarted = true;
+        }
+    }
+
+    function setContractUri(string memory uri_) public onlyOwner {
+        _contractUri = uri_;
+    }
+
+    function setProvenanceHash(string memory _provenanceHash)
+        external
+        onlyOwner
+        whenProvenanceNotFrozen
+    {
+        provenanceHash = _provenanceHash;
+    }
+
+    function setOffsetValue(uint256 _offsetValue) external onlyOwner {
+        require(!saleStarted, "sale already begun");
+        offsetValue = _offsetValue;
+    }
+
+    function setMaxPerMint(uint256 _maxPerMint) external onlyOwner {
+        maxPerMint = _maxPerMint;
+    }
+
+    function setMintPrice(uint256 _mintPrice) external onlyOwner {
+        mintPrice = _mintPrice;
+    }
+
+    function setBaseUri(string memory _baseUri)
+        external
+        onlyOwner
+        whenMetadataNotFrozen
+    {
+        baseUri = _baseUri;
+        emit SetBaseUri(baseUri);
+    }
+
+    function mintForCommunity(address _to, uint256 _numberOfTokens)
```

Comment:
Let's have a batch version of this, or have a separate multi-call contract. I recall it wasn't a nice experience using this for thousands of airdrops